### ISSUE TITLE
Set parameters on a per-topic basis

### ISSUE THREAD

**Issue by drmibell** [bartbutenaers/node-red-contrib-msg-resend#3]
This is a very useful node, and the recent improvements (programmatic input of values and resend per topic) make it even better. However, these two features interact in an odd way. I would have expected values like resend_max_count to operate on a per topic basis when the topic dependent option is in effect, but my testing seems to show that the value received most recently is used for all topics. Am I right, and if so is this a bug?

**Comment by bartbutenaers** [comment]
Hi Michael,

You are right that the **same resend_max_count** is currently being used for all topics.  But it is not really a bug since that was the original purpose.

But indeed I can imagine that it could be useful to have a **resend_max_count per topic** in some cases.  So I would like to implement your suggestion.  Was thinking about following approach:
- The config view of my node still contains the current resend_max_count field:
![image](https://user-images.githubusercontent.com/14224149/29748501-70c9b0c4-8b18-11e7-9c24-ea027408062b.png)  
This will be the _default_ value, in case certain topics don't specify their own value.  This way I can avoid to break existing flows.
- When a message arrives containing the `msg.resend_max_count` field, that value will be used to:
      + Update the single common default value when no `msg.topic` field is available
      + Update the topic specific value when a `msg.topic` field is available
- Perhaps it could be useful (when only few topics are being used) to allow users in the config view to specify a default resend_max_count value per topic (when topic dependent option is active):
![image](https://user-images.githubusercontent.com/14224149/29748630-59bff634-8b1a-11e7-8628-858b584545f1.png)

Please let me know if this is the kind of solution you are looking for...

Kind regards,
Bart

**Comment by drmibell** [comment]
Bart,

This is exactly the kind of solution I was looking for. I control several
devices (open loop) that need repeat commands for reliability, while others
do not. Since the target device is identified by the msg.topic, it would be
great to be able to use the same resend node in every case, just changing
msg.resend_max_count. Your second option, setting resend_max_count on a
per-topic basis in the configuration dialog, seems less useful -- as you
say, it really only makes sense for a very small number of topics.

At the risk of sounding greedy, would it be possible to do something
similar with resend_interval?

Thanks for your prompt response and willingness to help.

Best regards,
Mike

**Comment by bartbutenaers** [comment]
Hi Michael (@drmibell),

I have committed ***version 0.0.7***, which allows you to control the node for EACH topic separately, when you setup the resend node via an input message.  If the message contains a `msg.topic` property, the resend properties (`resend_interval`, `resend_max_count` and `resend_force_clone`) will be ***set only for the specified topic***.

A test flow:
![image](https://user-images.githubusercontent.com/14224149/29762418-aebfc326-8bce-11e7-846f-292109a286fd.png)
```
[{"id":"377fed0e.5c7f82","type":"msg-resend","z":"6beebf75.ed0b5","interval":"1","maximum":"5","bytopic":true,"clone":false,"name":"TopicResending","x":668.7656555175781,"y":2111.750325202942,"wires":[["6eac3895.95eab8"]]},{"id":"4d0c9d29.3a57d4","type":"inject","z":"6beebf75.ed0b5","name":"Test 1","topic":"","payload":"","payloadType":"str","repeat":"","crontab":"","once":false,"x":274.758731842041,"y":1949.5521125793457,"wires":[["4246d128.05801","7dcaa41c.33282c","dd1155e.c5313a8"]]},{"id":"6eac3895.95eab8","type":"debug","z":"6beebf75.ed0b5","name":"Resend test","active":true,"console":"false","complete":"payload","x":853.7656555175781,"y":2111.6635026931763,"wires":[]},{"id":"4246d128.05801","type":"change","z":"6beebf75.ed0b5","name":"Set properties","rules":[{"t":"set","p":"resend_interval","pt":"msg","to":"2","tot":"num"},{"t":"set","p":"resend_max_count","pt":"msg","to":"4","tot":"num"},{"t":"set","p":"topic","pt":"msg","to":"topic1","tot":"str"},{"t":"set","p":"payload","pt":"msg","to":"FromTopic1","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":454.0039405822754,"y":1949.859577178955,"wires":[["377fed0e.5c7f82"]]},{"id":"7dcaa41c.33282c","type":"change","z":"6beebf75.ed0b5","name":"Set properties","rules":[{"t":"set","p":"resend_interval","pt":"msg","to":"3","tot":"num"},{"t":"set","p":"resend_max_count","pt":"msg","to":"8","tot":"num"},{"t":"set","p":"topic","pt":"msg","to":"topic2","tot":"str"},{"t":"set","p":"payload","pt":"msg","to":"FromTopic2","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":451.76565170288086,"y":2032.0001726150513,"wires":[["377fed0e.5c7f82"]]},{"id":"c96d8b5.ff85478","type":"comment","z":"6beebf75.ed0b5","name":"1. TOPIC-DEPENDENT RESENDING","info":"","x":276,"y":1864.84765625,"wires":[]},{"id":"fd94db60.3df3e8","type":"comment","z":"6beebf75.ed0b5","name":"Topic 1 with interval of 2 sec and maximum of 4 messages","info":"","x":408.01171875,"y":1911.859375,"wires":[]},{"id":"369a2e23.5b2fa2","type":"comment","z":"6beebf75.ed0b5","name":"Topic 2 with interval of 3 sec and maximum of 8 messages","info":"","x":407.765625,"y":1999,"wires":[]},{"id":"b6686ba0.a4d9c8","type":"comment","z":"6beebf75.ed0b5","name":"Topic 3 with default interval of 1 sec and maximum of 5 messages","info":"","x":426.1509895324707,"y":2078.5171909332275,"wires":[]},{"id":"5075431.8980bbc","type":"msg-resend","z":"6beebf75.ed0b5","interval":"1","maximum":"5","bytopic":false,"clone":false,"name":"TopicResending","x":662.765625,"y":2265.75,"wires":[["6eac3895.95eab8"]]},{"id":"190a67c2.b4d9e8","type":"inject","z":"6beebf75.ed0b5","name":"Test 2","topic":"","payload":"WithoutTopic","payloadType":"str","repeat":"","crontab":"","once":false,"x":268.7587013244629,"y":2265.551787376404,"wires":[["b501d3dd.5e0f9"]]},{"id":"b501d3dd.5e0f9","type":"change","z":"6beebf75.ed0b5","name":"Set properties","rules":[{"t":"set","p":"resend_interval","pt":"msg","to":"4","tot":"num"},{"t":"set","p":"resend_max_count","pt":"msg","to":"3","tot":"num"}],"action":"","property":"","from":"","to":"","reg":false,"x":448.00391006469727,"y":2265.859251976013,"wires":[["5075431.8980bbc"]]},{"id":"1c6e6c37.bb13c4","type":"comment","z":"6beebf75.ed0b5","name":"2. TOPIC-INDEPENDENT RESENDING","info":"","x":269.9999694824219,"y":2180.847331047058,"wires":[]},{"id":"610128d2.f3dbb8","type":"comment","z":"6beebf75.ed0b5","name":"Without topic with interval of 4 sec and maximum of 3 messages","info":"","x":422.0116882324219,"y":2227.859049797058,"wires":[]},{"id":"3f5546cf.1a117a","type":"msg-resend","z":"6beebf75.ed0b5","interval":"1","maximum":"5","bytopic":true,"clone":false,"name":"TopicResending","x":661.765625,"y":2409.75,"wires":[["6eac3895.95eab8"]]},{"id":"571e837d.eb857c","type":"inject","z":"6beebf75.ed0b5","name":"Test 3","topic":"topic4","payload":"MessageThatIsNotResended","payloadType":"str","repeat":"","crontab":"","once":false,"x":267.7587013244629,"y":2410.551787376404,"wires":[["16ef7606.3c328a"]]},{"id":"16ef7606.3c328a","type":"change","z":"6beebf75.ed0b5","name":"Set properties","rules":[{"t":"set","p":"resend_interval","pt":"msg","to":"5","tot":"num"},{"t":"set","p":"resend_max_count","pt":"msg","to":"2","tot":"num"},{"t":"set","p":"resend_ignore","pt":"msg","to":"true","tot":"bool"}],"action":"","property":"","from":"","to":"","reg":false,"x":448.00391006469727,"y":2409.859251976013,"wires":[["3f5546cf.1a117a","6a4b1ffb.2869"]]},{"id":"285e3f52.5a82","type":"comment","z":"6beebf75.ed0b5","name":"3. SET PROPERTIES (BY MSG THAT SHOULD BE IGNORED)","info":"","x":348.9999694824219,"y":2329.847331047058,"wires":[]},{"id":"9472bb4.2d21c48","type":"comment","z":"6beebf75.ed0b5","name":"Setup topic 4 : interval to 5 sec and maximum of 2 messages","info":"","x":411.0116882324219,"y":2372.859049797058,"wires":[]},{"id":"5b9a561c.435a48","type":"comment","z":"6beebf75.ed0b5","name":"Resend topic 4 with the updated settings","info":"","x":382.0186309814453,"y":2458.0574655532837,"wires":[]},{"id":"dd1155e.c5313a8","type":"change","z":"6beebf75.ed0b5","name":"Set properties","rules":[{"t":"set","p":"topic","pt":"msg","to":"topic3","tot":"str"},{"t":"set","p":"payload","pt":"msg","to":"FromTopic3","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":450.765625,"y":2111.75,"wires":[["377fed0e.5c7f82"]]},{"id":"6a4b1ffb.2869","type":"change","z":"6beebf75.ed0b5","name":"Set properties","rules":[{"t":"set","p":"topic","pt":"msg","to":"topic4","tot":"str"},{"t":"set","p":"payload","pt":"msg","to":"FromTopic4","tot":"str"},{"t":"delete","p":"resend_interval","pt":"msg"},{"t":"delete","p":"resend_max_count","pt":"msg"},{"t":"delete","p":"resend_ignore","pt":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":484.7656478881836,"y":2496.750093460083,"wires":[["3f5546cf.1a117a"]]}]
```
I have not published this new version on NPM yet, since it has been tested further.  You can test this new version 0.0.7 by installing it directly from github (in your Node-Red folder):
`npm install https://github.com/bartbutenaers/node-red-contrib-msg-resend.git`

Hey Jason (@muglar) and Nick (@elyobelyob): I tried to test as much scenarios as possible, but I would really appreciate if you guys could also give me feedback whether this new version doesn't break your existing flows.  Thanks a lot !!!!

Kind regards,
Bart

**Comment by bartbutenaers** [comment]
Hi Michael (@drmibell),,
Have you ever tested the functionality that you had requested?
Bart

**Comment by drmibell** [comment]
Bart,
Sorry, I missed your message regarding version 0.0.7. I tried version 0.0.6
and found the per-topic control worked as described. Unfortunately, I could
not use it my application because of the high input rate error. (I
sometimes need to be able to cancel an input and its repeats immediately.)
It looks like the latest version will solve this, so I will download it and
let you know. Thanks for your help.
Regards,
Mike


On 11 February 2018 at 14:19:21 , bartbutenaers (notifications@github.com)
wrote:

Hi Michael (@drmibell <https://github.com/drmibell>),,
Have you ever tested the functionality that you had requested?
Bart

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/bartbutenaers/node-red-contrib-msg-resend/issues/3#issuecomment-364779132>,
or mute the thread

**Comment by drmibell** [comment]
Hi Bart,
Version 0.0.7 seems to be doing a great job. It turns out that I also need the Output Count feature in my application, so your changes are very timely. Thanks again for a nice bit of work.
Regards,
Mike

**Comment by bartbutenaers** [comment]
Hey Mike,

Thanks for the feedback!  
I have published version **0.0.7**.  This means you can install it again using the normal npm commands. 

Kind regards,
Bart
